### PR TITLE
CR #21180 Camera control app

### PR DESCRIPTION
**Pull request by humbletim**
This PR implements a Camera Movement App that lets you configure easing parameters to smooth-out Avatar movements and turning, helping to avoid certain kinds of jitter when filming in an Avatar-as-a-Camera scenario.  Worklist item: https://worklist.net/21180

* *Note: this app is intended only for Desktop mode -- previous discussions/specs indicated that some kind of warning should be displayed if someone attempts use in HMD (TBD?)*
*  *the "Avatars snap look at camera" and "Enable minimal mouse cursor" checkboxes depend on PR#10625 -- so are only fully-testable if running that build.  The rest of the app / settings can be tested using the most recent beta release.*

### Testing
* Load the `app-camera-move.js` Client script from the `unpublishedScripts/marketplace/camera-move` staging area; here's a rawgit link to use:
[unpublishedScripts/marketplace/camera-move/app-camera-move.js](https://cdn.rawgit.com/humbletim/hifi/c9874a7/unpublishedScripts/marketplace/camera-move/app-camera-move.js)
* Verify the new "Cam Drive" button appears in the toolbar.
* Press <kbd>space</kbd> to activate Camera Move Mode.
* Verify the app icon becomes "active".
* Verify mouse looking and keyboard movements now have smoothing applied.
* Experiment with various slider positions and values -- verifying that in general the changes get applied to avatar movements.
* Observe the current state of the "Enable snap turn in HMD" checkbox.
    * Open "Settings > Avatar" and change the "Snap turn when in HMD" setting.
    * Back in the Camera Movement App, verify the "Enable snap turn in HMD" checkbox reflects the new state.
* Press <kbd>space</kbd> to de-active Camera Move Mode.
* Verify avatar movements/turning take place normally again (ie: without any smoothing applied).

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/255/HighFidelity-Beta-PR10574-6f455d8429536d7a2ef80bf3c68e1a9654698d0b.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/255/HighFidelity-Beta-PR10574-6f455d8429536d7a2ef80bf3c68e1a9654698d0b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/273/HighFidelity-Beta-PR10574-4d7999c2ceed477bcf6f271bc396eb1330aa7485.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/273/HighFidelity-Beta-PR10574-4d7999c2ceed477bcf6f271bc396eb1330aa7485.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/289/HighFidelity-Beta-PR10574-c06d50b28c4ac2151c9a7503acd56c3ac756a062.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/289/HighFidelity-Beta-PR10574-c06d50b28c4ac2151c9a7503acd56c3ac756a062.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/537/HighFidelity-Beta-PR10574-16170cd754c1568c6c7acd8b75eaaff8af049599.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/537/HighFidelity-Beta-PR10574-16170cd754c1568c6c7acd8b75eaaff8af049599.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/571/HighFidelity-Beta-PR10574-c8d503f717f430abdc96ee40337aa325fbd42453.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/571/HighFidelity-Beta-PR10574-c8d503f717f430abdc96ee40337aa325fbd42453.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/799/HighFidelity-Beta-PR10574-2a8b91e45e9781b7f74bdb26f29e4e019a883909.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/799/HighFidelity-Beta-PR10574-2a8b91e45e9781b7f74bdb26f29e4e019a883909.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/953/HighFidelity-Beta-PR10574-01436675b113c4b96128cbd0e05be7c5e40e8629.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/953/HighFidelity-Beta-PR10574-01436675b113c4b96128cbd0e05be7c5e40e8629.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/954/HighFidelity-Beta-PR10574-9d4f9b0a4b0a69a6603fffc7b5674e4d97c64790.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/954/HighFidelity-Beta-PR10574-9d4f9b0a4b0a69a6603fffc7b5674e4d97c64790.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/956/HighFidelity-Beta-PR10574-c9874a748972a99da1f3574c95f8cfd1a52afbb2.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/956/HighFidelity-Beta-PR10574-c9874a748972a99da1f3574c95f8cfd1a52afbb2.dmg

**Comment by Nex-Pro** [comment]
Checking this.

**Comment by Nex-Pro** [comment]
Still in progress.

**Comment by Nex-Pro** [comment]
@humbletim 

Results so far.

CR #21180 Camera control app #10574

Tested on win10 i7 and GTX 1070


Load the app-camera-move.js Client script from the unpublishedScripts/marketplace/camera-move staging area; here's a rawgit link to use:

unpublishedScripts/marketplace/camera-move/app-camera-move.js **DONE**


Verify the new "Cam Drive" button appears in the toolbar. **SUCCESS**

Press space to activate Camera Move Mode. **DONE**

Verify the app icon becomes "active". **SUCCESS**

Verify mouse looking and keyboard movements now have smoothing applied. **Works but jerky**

Experiment with various slider positions and values -- verifying that in general the changes get applied to avatar movements.

Observe the current state of the "Enable snap turn in HMD" checkbox.

Open "Settings > Avatar" and change the "Snap turn when in HMD" setting.

Back in the Camera Movement App, verify the "Enable snap turn in HMD" checkbox reflects the new state.
Press space to de-active Camera Move Mode.

Verify avatar movements/turning take place normally again (ie: without any smoothing applied).


Feedback/Observations:

Tim, awesome as ever.

This is a nice addition to the system.

It is so complex it warrants a user manual or a set of tutorial pages/videos.

Providing an option to deslect avatar rendering is a good and necessary feature.

Nice to be able to move (intentionally) at high speeds.



### Ease-in:
    
    At a setting of 300Ms it takes a long time to accelerate to the max speed even when setting the ease-in to 10.0.
    It is possible to set the translation ease-in to numbers much higher than 10 but the slider stops at 10.
    Perhaps a log slider might be a better design.
    Ease in can take large inputs 9999999 and seems to use the high figures.
    However, closing the dialog and re-opening displays the ease-in at 10.0 but retains the higher setting in use.
    Setting 10.00 when 10.00 is displayed in this situation does not change the actual ease-in. Changing to zero then setting 10.00 does!
    There appears to be some repeatable hesitation in the easing when using the forward and back keys. Max Linear velocity at 100 ease-in at 0.28.
    This is less noticeable when 'Update continuously' is set in Timing. This applies to all three update modes. Setting 'Use constant delta time' does not improve things.
    Does this point to an issue with keyboard polling ?


### Ease out:

    Ease-out co-efficient is capped at 10.00


### Max Linear Velocity:

    Max Linear Velocity can be set to huge values but speed in interface seems to be capped to 500Ms


### Max angular velocity:

    There seems too be no cap on Max angular velocity :)


### Eye vector rendering:

    Eye vectors are only rendering in front of entities. They are not rendering against skyboxes.


### Prevent Roll:

    All kinds of wierdness ensues. ( I felt the need to get technical here. )


### Tooltips:

    Tooltips for Keyboard and mouse multipliers could be more informative.

    Tips for Update modes aren't very informative for the average user. 


### Timing:
    
    Update @ FPS defaults to 90 but desktop FPS is normally 60. Is this intended ?

File: unpublishedScripts/marketplace/camera-move/modules/movement-utils.js [hunk 1]
```diff
@@ -0,0 +1,773 @@
...
+
+        log('EXCLUSIONS:' + exclude);
+
+        template.channels = template.channels.filter(_utils.bind(this, function(item, i) {
+            debugPrint('channel['+i+']', item.from && item.from.makeAxis, item.to, JSON.stringify(item.filters) || '');
+            // var hasFilters = Array.isArray(item.filters) && !item.filters[1];
+            item.filters = Array.isArray(item.filters) ? item.filters :
+                typeof item.filters === 'string' ? [ { type: item.filters }] : [ item.filters ];
+
+            if (/Mouse/.test(item.from && item.from.makeAxis)) {
```
Comment (Atlante45) [inline unpublishedScripts/marketplace/camera-move/modules/movement-utils.js]: I've seen several instances of this, can you explain what you're doing here?
It looks like you testing a boolean against a regex.

Comment (Atlante45) [inline unpublishedScripts/marketplace/camera-move/modules/movement-utils.js]: Nvm, I got it. `&&` returns the last value if everything before it is true.

**Pull request merged**